src/control/server/iosrv.go

```diff
@@ -506,13 +506,15 @@ func mgmtJoin(ap string, tc *security.TransportConfig, req *mgmtpb.JoinReq) (*mg
 
 	client := mgmtpb.NewMgmtSvcClient(conn)
 
-	resp, err := client.Join(context.Background(), req)
-	if err != nil {
-		return nil, errors.Wrapf(err, "join %s %v", ap, *req)
-	}
-	if resp.Status != 0 {
-		return nil, errors.Errorf("join %s %v: %d\n", ap, *req, resp.Status)
+	for {
+		resp, err := client.Join(context.Background(), req)
+		if err == nil {
```

Comment:
It would be preferable at a minimum to add a logging statement here (with Errorf, I guess) so that the admin has some idea of what might be going on if it gets stuck in a retry loop. Even better I think would be to just define a maximum retry value (5? 10?) and return an error if it's hit. I understand that there will be more work on this area in the future, but I don't like the idea of adding a possible infinite loop which will hopefully get fixed later.

Reply:
I agree with not wanting an infinite loop.  Logging an info message after each retry, with an error message and error return after a maximum retry time interval (whatever you decide) would be helpful.

Reply:
I have added some log.Debugf calls and (hopefully) proper context.Context support. You may add a timeout configuration parameter and implement it using this context.Context argument.

That said, one daos_server is only one small piece of a DAOS system. If I'd like to start my 2000 servers simultaneously, many factors may cause the server running the MS replicas to become ready later than some other servers. The pdsh fan out configuration if I'm using pdsh, systemd configuration if I'm using systemd, network situations, whether the servers run on VMs, etc. Similar to Lustre, daos_servers would be easier to manage if they just take their to retry, synchronize, and eventually work things out, regardless of when each one becomes ready, without terminating waiting for me to find and restart the terminated ones. I therefore consider retrying indefinitely a reasonable default.

Hope this helps. Thanks.

Reply:
That does help, thank you. I agree in that context that it does probably make sense to just retry indefinitely. My concern is mostly about making sure that the admin is not surprised. If we document that the expected behavior is to start up and try forever to join, that's fine.

Another thought is that this behavior might be configurable based on the intended usage. We might want to add an optional timeout value that allows the admin (or us in our test environment) to specify that we expect all members to have joined by some time. We can add that in a future commit, though.

Reply:
> Another thought is that this behavior might be configurable based on the intended usage. We might want to add an optional timeout value that allows the admin (or us in our test environment) to specify that we expect all members to have joined by some time. We can add that in a future commit, though.

Yes, this is what I was trying to say above: "You may add a timeout configuration parameter later and implement it using this context.Context argument."

Reply:
E.g., I tried passing context.WithTimeout(context.Background(), 15*time.Second) to mgmtJoin.